### ISSUE TITLE
Attributes leak scalars

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12379]
Migrated from [rt.perl.org#114764](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114764) (status was 'resolved')

Searchable as RT114764$

**Comment by p5pRT** [comment]
### From @jdhedden

#### Created by @jdhedden



The use of attributes causes leaked scalars&#8203;:

> perl \-MTest&#8203;::LeakTrace&#8203;::Script \-l \-e 'my $x'
\(No output \- nothing leaked\.\)

> perl \-MTest&#8203;::LeakTrace&#8203;::Script \-l \-e 'my $x : shared'
leaked SCALAR\(0x2003bf00\) from \-e line 1\.
leaked HASH\(0x2006aca8\) from /usr/lib/perl5/5\.16\.1/cygwin/attributes\.pm line 59\.
leaked ARRAY\(0x2006ac18\) from
/usr/lib/perl5/5\.16\.1/cygwin/attributes\.pm line 59\.
leaked SCALAR\(0x2006ac00\) from
/usr/lib/perl5/5\.16\.1/cygwin/attributes\.pm line 59\.
leaked GLOB\(0x2006abe8\) from /usr/lib/perl5/5\.16\.1/cygwin/attributes\.pm line 59\.

\(Note that this is not a 'threads&#8203;::shared' bug as neither 'threads' nor
'threads&#8203;::shared' is being loaded\.\)

Curiously\, changing 'my' to 'our' doesn't leak&#8203;:

> perl \-MTest&#8203;::LeakTrace&#8203;::Script \-l \-e 'our $x : shared'

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.16.1:

Configured by Jerry at Tue Aug 14 12:59:06 EDT 2012.

Summary of my perl5 (revision 5 version 16 subversion 1 patch 53802)
configuration:
  Snapshot of: 06d742c02d396d6515581002f376b55ab1972c1b
  Platform:
    osname=cygwin, osvers=1.7.16(0.26253), archname=cygwin-thread-multi-64int
    uname='cygwin_nt-5.1 med-heddenj 1.7.16(0.26253) 2012-07-20 22:55
i686 cygwin '
    config_args='-de -Duse64bitint -Dusethreads -Uusemymalloc
-Dinc_version_list=none -Dnoextensions=DB_File Devel/DProf Devel/Peek
GDBM_File IPC/SysV NDBM_File ODBM_File Sys/Syslog Text/Soundex
Time/Piece attrs B/Debug B/Lint -A append:ccflags= -DNO_MATHOMS -A
define:optimize=-Os -pipe -funit-at-a-time -march=pentium4
-mfpmath=sse -mieee-fp -mmmx -msse -msse2'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__
-DNO_MATHOMS -fno-strict-aliasing -pipe -fstack-protector
-I/usr/local/include',
    optimize='-Os -pipe -funit-at-a-time -march=pentium4 -mfpmath=sse
-mieee-fp -mmmx -msse -msse2',
    cppflags='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -DNO_MATHOMS
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.5.3', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8,
Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='g++', ldflags =' -Wl,--enable-auto-import
-Wl,--export-all-symbols -Wl,--enable-auto-image-base -s
-fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /lib
    libs=-ldb -ldl -lcrypt
    perllibs=-ldl -lcrypt
    libc=/usr/lib/libc.a, so=dll, useshrplib=true, libperl=cygperl5_16_1.dll
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' -s'
    cccdlflags=' ', lddlflags=' --shared  -Wl,--enable-auto-import
-Wl,--export-all-symbols -Wl,--enable-auto-image-base -s
-L/usr/local/lib -fstack-protector'

Locally applied patches:



@INC for perl 5.16.1:
    /usr/lib/perl5/site_perl/5.16.1/cygwin
    /usr/lib/perl5/site_perl/5.16.1
    /usr/lib/perl5/5.16.1/cygwin
    /usr/lib/perl5/5.16.1
    .


Environment for perl 5.16.1:
    CYGWIN=nodosfilewarning
    HOME=/home/heddenj
    LANG=C
    LANGUAGE=C
    LC_ALL=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/heddenj/bin:/usr/local/src/perl/bin:/usr/local/bin:/usr/bin:/c/Progra~1/Java/jre6/bin:/c/WINDOWS/system32:/c/WINDOWS:/c/WINDOWS/System32/Wbem:/c/Program
Files/ATI Technologies/ATI.ACE/Core-Static:/c/Program
Files/TortoiseSVN/bin:/c/PYTHON27:/c/Program Files/MySQL/MySQL Server
5.5/bin:/c/WINDOWS/system32/WindowsPowerShell/v1.0:.
    PERLIO=perlio
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed Sep 05 09&#8203;:05&#8203;:12 2012\, jdhedden@&#8203;cpan\.org wrote&#8203;:
> This is a bug report for perl from jdhedden@&#8203;cpan\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> The use of attributes causes leaked scalars&#8203;:
> 
> > perl \-MTest&#8203;::LeakTrace&#8203;::Script \-l \-e 'my $x'
> \(No output \- nothing leaked\.\)
> 
> > perl \-MTest&#8203;::LeakTrace&#8203;::Script \-l \-e 'my $x : shared'
> leaked SCALAR\(0x2003bf00\) from \-e line 1\.
> leaked HASH\(0x2006aca8\) from
> /usr/lib/perl5/5\.16\.1/cygwin/attributes\.pm line 59\.
> leaked ARRAY\(0x2006ac18\) from
> /usr/lib/perl5/5\.16\.1/cygwin/attributes\.pm line 59\.
> leaked SCALAR\(0x2006ac00\) from
> /usr/lib/perl5/5\.16\.1/cygwin/attributes\.pm line 59\.
> leaked GLOB\(0x2006abe8\) from
> /usr/lib/perl5/5\.16\.1/cygwin/attributes\.pm line 59\.
> 
> \(Note that this is not a 'threads&#8203;::shared' bug as neither 'threads'
> nor
> 'threads&#8203;::shared' is being loaded\.\)
> 
> Curiously\, changing 'my' to 'our' doesn't leak&#8203;:
> 
> > perl \-MTest&#8203;::LeakTrace&#8203;::Script \-l \-e 'our $x : shared'
> 
> 

This excerpt from 'perldoc attributes' may be relevant to this problem&#8203;:

\#\#\#\#\#
WARNING&#8203;: attribute declarations for variables are still evolving\.
The semantics and interfaces of such declarations could change in
future versions\.  They are present for purposes of experimentation
with what the semantics ought to be\.  Do not rely on the current
implementation of this feature\.

\.\.\.

The setting of subroutine attributes happens at compile time\.
Variable attributes in C\<our> declarations are also applied at compile time\.
However\, C\<my> variables get their attributes applied at run\-time\.
This means that you have to I\<reach> the run\-time component of the C\<my>
before those attributes will get applied\.
\#\#\#\#\#

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jdhedden

On Wed\, Sep 5\, 2012 at 9&#8203;:22 PM\, James E Keenan via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> This excerpt from 'perldoc attributes' may be relevant to this problem&#8203;:
> \#\#\#\#\#
> WARNING&#8203;: attribute declarations for variables are still evolving\.
> The semantics and interfaces of such declarations could change in
> future versions\.  They are present for purposes of experimentation
> with what the semantics ought to be\.  Do not rely on the current
> implementation of this feature\.

Useful information\, but I don't see how it's relevant\.  There's a bug in the
current implementation \(leaking scalars\)\, and that is what is being reported\.

As far as the 'experimentation' qualifier\, that's a bit out\-dated\.  For
example\, the '&#8203;:shared' attribute if very much locked in\, in my opinion\.

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed by 9fa29fa7929\.

\-\- 

Father Chrysostomos

s||$\{s/\.\*/|;
/s\}Just another Perl hacker\,
print

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Fixed by 9fa29fa7929\.

\-\- 

Father Chrysostomos

s||$\{s/\.\*/|;
/s\}Just another Perl hacker\,
print

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
